sale_pricelist_triple_discount/model/sale.py

```diff
@@ -1,55 +1,192 @@
+                    for t in taxes:
+                        if t["id"] == tax.id or t["id"] in tax.children_tax_ids.ids:
+                            res[group]["amount"] += t["amount"]
+                            res[group]["base"] += t["base"]
+            res = sorted(res.items(), key=lambda l: l[0].sequence)
+            order.amount_by_group = [
+                (
+                    entry[0].name,
+                    entry[1]["amount"],
+                    entry[1]["base"],
+                    fmt(entry[1]["amount"]),
+                    fmt(entry[1]["base"]),
+                    len(res),
+                )
+                for entry in res
+            ]
 
 
 class SaleOrderLine(models.Model):
     _inherit = "sale.order.line"
 
-    @api.multi
-    def product_id_change(
-        self,
-        pricelist,
-        product,
-        qty=0,
-        uom=False,
-        qty_uos=0,
-        uos=False,
-        name="",
-        partner_id=False,
-        lang=False,
-        update_tax=True,
-        date_order=False,
-        packaging=False,
-        fiscal_position=False,
-        flag=False,
-    ):
-        res = super(SaleOrderLine, self).product_id_change(
-            pricelist,
-            product,
-            qty=qty,
-            uom=uom,
-            qty_uos=qty_uos,
-            uos=uos,
-            name=name,
-            partner_id=partner_id,
-            lang=lang,
-            update_tax=update_tax,
-            date_order=date_order,
-            packaging=packaging,
-            fiscal_position=fiscal_position,
-            flag=flag,
+    @api.onchange(
+        "product_id", "price_unit", "product_uom", "product_uom_qty", "tax_id"
+    )
+    def _onchange_discount(self):
+        super()._onchange_discount()
+
+        pricelist = self.order_id.pricelist_id
+        if not (
+            self.product_id
+            and self.product_uom
+            and self.order_id.partner_id
+            and pricelist
+            and pricelist.discount_policy == "without_discount"
+            and self.env.user.has_group("product.group_discount_per_so_line")
+        ):
+            return
+
+        current_pricelist = self.order_id.pricelist_id
+        list_price = current_pricelist.price_rule_get(
+            self.product_id.id, self.product_uom_qty or 1.0, self.order_id.partner_id.id
+        )
+        rule_id = (
+            list_price.get(current_pricelist.id)
+            and list_price[current_pricelist.id][1]
+            or False
         )
-        if product and pricelist:
-            value = res["value"]
-            current_pricelist = self.env["product.pricelist"].browse(pricelist)
+        if rule_id:
+            rule = self.env["product.pricelist.item"].browse(rule_id)
+            read_rule = rule.read(["discount2", "discount3"])[0]
+            self.discount2 = read_rule["discount2"] or 0.00
+            self.discount3 = read_rule["discount3"] or 0.00
+
+    @api.depends(
```

Comment:
It is safer to only add the new fields you need to depend on, it works because `extend` is used in https://github.com/odoo/odoo/blob/4fd3b7a062e9f8107587ceed98ebb2f51627a4d6/odoo/fields.py#L452